itou/metabase/management/commands/_utils.py

```diff
@@ -0,0 +1,135 @@
+
+
+def convert_boolean_to_int(b):
+    # True => 1, False => 0, None => None.
+    return None if b is None else int(b)
+
+
+def compose(f, g):
+    # Compose two lambda methods.
+    # https://stackoverflow.com/questions/16739290/composing-functions-in-python
+    # I had to use this to solve a cryptic
+    # `RecursionError: maximum recursion depth exceeded` error
+    # when composing convert_boolean_to_int and c["lambda"].
+    return lambda *a, **kw: f(g(*a, **kw))
+
+
+def get_choice(choices, key):
+    choices = dict(choices)
+    # Gettext fixes `can't adapt type '__proxy__'` error
+    # due to laxy_gettext and psycopg2 not going well together.
+    # See https://code.djangoproject.com/ticket/13965
+    if key in choices:
+        return gettext(choices[key])
+    return None
+
+
+def chunks(items, n):
+    """
+    Yield successive n-sized chunks from items.
+    """
+    for i in range(0, len(items), n):
+        yield items[i : i + n]
+
+
+def anonymize(value, salt):
+    """
+    Use a salted hash to anonymize sensitive ids,
+    mainly job_seeker id and job_application id.
+    """
+    return salted_hmac(salt, value, secret=settings.SECRET_KEY).hexdigest()
+
+
+def get_first_membership_join_date(memberships):
+    memberships = list(memberships.all())
+    # We have to do all this in python to benefit from prefetch_related.
+    if len(memberships) >= 1:
+        memberships.sort(key=lambda o: o.joined_at)
+        first_membership = memberships[0]
+        return first_membership.joined_at
+    return None
+
+
+def _get_job_seeker_id_to_hiring_siae():
+    """
+    Ideally the job_seeker would have a unique hiring so that we can
+    properly link the approval back to the siae. However we already
+    have many job_seekers with two or more hirings. In this case
+    we consider the latest hiring, which is an ugly workaround
+    around the issue that we do not have a proper approval=>siae
+    link yet.
+
+    Preload this association data for best performance.
+    """
+    # Order by created_at so that most recent hiring overrides older ones.
+    hirings = (
+        JobApplication.objects.filter(state=JobApplicationWorkflow.STATE_ACCEPTED)
+        .order_by("created_at")
+        .select_related("to_siae")
+    )
+    job_seeker_id_to_hiring_siae = {}
+    for hiring in hirings:
+        job_seeker_id = hiring.job_seeker_id
+        job_seeker_id_to_hiring_siae[job_seeker_id] = hiring.to_siae
+    return job_seeker_id_to_hiring_siae
+
+
+JOB_SEEKER_ID_TO_HIRING_SIAE = _get_job_seeker_id_to_hiring_siae()
+
+
+def get_department_and_region_columns(name_suffix="", comment_suffix="", custom_lambda=lambda o: o):
+    return [
+        {
+            "name": f"département{name_suffix}",
+            "type": "varchar",
+            "comment": f"Département{comment_suffix}",
+            "lambda": lambda o: custom_lambda(o).department if custom_lambda(o) else None,
```

Comment:
Mon 1er instinct en lisant ça est de me dire, hum, une _lambda_ dans une _lambda_ si une _lambda_… _there must be a better way_…

Reply:
De mémoire j'avais cherché sans trouver... pas facile de garder un one liner. Mais j'ai des nouvelles pistes dans https://stackoverflow.com/questions/2428557/concise-way-to-getattr-and-use-it-if-not-none-in-python

Reply:
Ce sera `getattr(custom_lambda(o), "department", None)` mais c'est un peu moche car ça ne lancera aucune exception si jamais je me suis planté et que l'attribut `departement` n'existe pas.